build v4.5.3 (with dynamic libraries)

grlee77 opened the pull request:
I cleaned up conda-forge/staged-recipes#682 which builds shared SuiteSparse libraries compatible with the  open [scikit-umfpack PR](onda-forge/staged-recipes#683).  Sstatic libraries actually get built as well, but `make install` doesn't install them. 

I am now using SuiteSparse_config.mk as in master, but it no longer seems necessary to have separate files for linux and OS X.

I changed the license info in meta.yaml slightly.  I moved LGPL v2 first as that is what the majority of the code seems to be licensed under (I only found one library with a BSD 3-clause license).  In the v4.5.3 source the licenses are more clearly stated in the separate /Doc/License.txt within each library's subfolder.

Adding cmake to the build dependencies was needed for building the included Metis v5.1.0.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

jakirkham commented:
Is there some reason not to install both static and dynamic libraries?

jakirkham commented:
> I am now using SuiteSparse_config.mk as in master, but it no longer seems necessary to have separate files for linux and OS X.

Yeah, I started to make them pretty similar at the end. For the most part, started using a bunch of `?=` instead of `=` allowed most of the configuration to happen in `build.sh`, which seemed preferable anyways. I emailed Tim Davis making this exact suggestion though I don't know if he will go for it.

jakirkham commented:
> I changed the license info in meta.yaml slightly. I moved LGPL v2 first as that is what the majority of the code seems to be licensed under (I only found one library with a BSD 3-clause license). In the v4.5.3 source the licenses are more clearly stated in the separate /Doc/License.txt within each library's subfolder.

Yeah, that seems reasonable to me. Licensing here is a bit of a pain unfortunately. Appreciate that you added Apache 2.0 for METIS.

jakirkham commented:
> Adding cmake to the build dependencies was needed for building the included Metis v5.1.0.

👍

jakirkham commented on recipe/SPQR_TBB.patch:
```diff
@@ -0,0 +1,11 @@
+ 
+ # SPQR depends on CHOLMOD, AMD, COLAMD, LAPACK, the BLAS and SuiteSparse_config
+-LDLIBS += -lamd -lcolamd -lcholmod -lsuitesparseconfig $(LAPACK) $(BLAS)
++LDLIBS += -lamd -lcolamd -lcholmod -lsuitesparseconfig $(LAPACK) $(BLAS) $(TBB)
```
Are we sure this patch is needed? It seemed to pick up TBB fine in 4.4.6 and 4.4.5. If this is indeed required, it might be good to drop Tim Davis a line about this regression.

grlee77 replied:
I didn't need it on OS X, but for some reason when I tried locally on my linux machine I was getting errors related to TBB.  I can try removing it here and see what happens.

jakirkham replied:
We're no longer using this patch, right? Maybe we should remove it.

jakirkham commented on recipe/build.sh:
```diff
@@ -19,5 +21,20 @@ export INSTALL_INCLUDE="${PREFIX}/include"
+if [ "$(uname)" == "Darwin" ]; then
+
+  # change link to METIS for CHOLMOD to reflect install path
+  install_name_tool -change $SRC_DIR/lib/libmetis.dylib @rpath/libmetis.dylib $PREFIX/lib/libcholmod.dylib
```
Why is this needed? Doesn't `conda-build` fix this anyways? If not, do we need to do something similar on Linux with `patchelf`?

grlee77 replied:
We can try removing it later.  I found that when I tried scikit-umfpack on OS X I got errors relating to the library not being found and this fixed it.

jakirkham commented on recipe/build.sh:
```diff
@@ -19,5 +21,20 @@ export INSTALL_INCLUDE="${PREFIX}/include"
 eval ${LIBRARY_SEARCH_VAR}="${PREFIX}/lib" make -j1
+
+# make install below fails to link libmetis unless it is copied to $PREFIX/lib
+cp $SRC_DIR/lib/libmetis$DYNAMIC_EXT $PREFIX/lib
```
Maybe we could just build METIS first and install it. Then build everything else and install them. It might save us a few ugly hacks that cause us problems later.

grlee77 replied:
Do you mean as a separate feedstock or as a separate step in build.sh here?

jakirkham replied:
As part of `build.sh`.

grlee77 replied:
Here is the relevant METIS section of the SuiteSparse Makefile
```
# configure and compile METIS, placing the libmetis.* library in
# SuiteSparse/lib and the metis.h include file in SuiteSparse/include.
metis: include/metis.h

# Install the shared version of METIS in SuiteSparse/lib.
# The SO_INSTALL_NAME commmand is only needed on the Mac, so *.dylib is
# hardcoded below.
include/metis.h:
ifeq (,$(MY_METIS_LIB))
	- ( cd metis-5.1.0 && $(MAKE) config shared=1 prefix=$(SUITESPARSE) )
	- ( cd metis-5.1.0 && $(MAKE) )
	- ( cd metis-5.1.0 && $(MAKE) install )
	- $(SO_INSTALL_NAME) $(SUITESPARSE)/lib/libmetis.dylib \
                             $(SUITESPARSE)/lib/libmetis.dylib
else
	@echo 'Using pre-installed METIS 5.1.0 library at ' '[$(MY_METIS_LIB)]'
endif
```
I think what we would want is prefix=$PREFIX so that it ends up installed where we would like as opposed to $SUITESPARSE ($SUITESPARSE is the same as $SRC_DIR)

Also, it seems like a bug in the Makefile that .dylib is listed there when the extension would be .so on linux?

grlee77 commented:
As far as the static libraries, they do also get built, but I didn't see a way to get `make install` to install them.  

After the make completes, I can see static libraries in various subfolders, e.g. for CHOLMOD the static lib is at `${SRC_DIR}/CHOLMOD/Lib/libcholmod.a`.  Obviously we could manually cp them, but maybe there is a way to do it from the Makefile that I overlooked?

grlee77 commented:
CI is now passing. will try removing the patch and adding install for the static libs

jakirkham commented:
Seems there is an issue copying the static libs over. Here is part of the CircleCI [log]( https://circleci.com/gh/conda-forge/suitesparse-feedstock/10 ).

```
cp: cannot stat `/feedstock_root/build_artefacts/work/SuiteSparse/CHOLAMD/Lib/libcholamd.a': No such file or directory
Command failed: /bin/bash -x -e /recipe_root/build.sh
```

grlee77 commented:
I see there is a PR at staged-recipes for a METIS package as well.  How do we make sure the METIS compiled here doesn't end up conflicting with that one?

Apparently the SPQR_TBB.patch wasn't needed (not sure why I needed it to build successfully on my local linux system).

jakirkham commented on recipe/build.sh:
```diff
@@ -19,5 +21,46 @@ export INSTALL_INCLUDE="${PREFIX}/include"
+if [ "$(uname)" == "Darwin" ]
+then
+    # this is copied from the SuiteSparse Makefile
+    install_name_tool -id ${PREFIX}/lib/libmetis.dylib ${PREFIX}/lib/libmetis.dylib
```
Is this still needed now that we handle the build of METIS separately?

grlee77 replied:
not sure if the `install_name_tool` call is necessary.  I copied it verbatim from the SuiteSparse Makefile.  I can try without and see if maybe it gets handled for us by conda build.

jakirkham commented:
> I see there is a PR at staged-recipes for a METIS package as well. How do we make sure the METIS compiled here doesn't end up conflicting with that one?

I'm not sure how much we should push that one. My concern is the licensing with it is totally opaque. Also, we will want to move to SuiteSpare 4.5.x eventually, which includes a copy/version of METIS that has clear licensing. In any event, let's cross that bridge when we get there.

> Apparently the SPQR_TBB.patch wasn't needed (not sure why I needed it to build successfully on my local linux system).

Interesting. It does seem to be working. What version of Linux were you using?

<br>
Basically, this is looking really good. Thanks for doing this @grlee77. I'll try to find some time tomorrow or Friday to give it a closer look.

grlee77 commented:
I just noticed that the linux case is still passing `-fopenmp`.  I can fix that.

> Interesting. It does seem to be working. What version of Linux were you using?

Ubuntu 14.04.  I didn't look into it much, but the errors reported were:
```
spqr_parallel.o: In function `spqr_zippy<std::complex<double> >::~spqr_zippy()':
spqr_parallel.cpp:(.text._ZN10spqr_zippyISt7complexIdEED2Ev[_ZN10spqr_zippyISt7complexIdEED5Ev]+0x3): undefined reference to `vtable for tbb::task'
spqr_parallel.o: In function `spqr_zippy<double>::~spqr_zippy()':
spqr_parallel.cpp:(.text._ZN10spqr_zippyIdED2Ev[_ZN10spqr_zippyIdED5Ev]+0x3): undefined reference to `vtable for tbb::task'
spqr_parallel.o: In function `spqr_zippy<std::complex<double> >::~spqr_zippy()':
spqr_parallel.cpp:(.text._ZN10spqr_zippyISt7complexIdEED0Ev[_ZN10spqr_zippyISt7complexIdEED0Ev]+0x3): undefined reference to `vtable for tbb::task'
spqr_parallel.o: In function `spqr_zippy<double>::~spqr_zippy()':
spqr_parallel.cpp:(.text._ZN10spqr_zippyIdED0Ev[_ZN10spqr_zippyIdED0Ev]+0x3): undefined reference to `vtable for tbb::task'
spqr_parallel.o: In function `spqr_zippy<std::complex<double> >::execute()':
spqr_parallel.cpp:(.text._ZN10spqr_zippyISt7complexIdEE7executeEv[_ZN10spqr_zippyISt7complexIdEE7executeEv]+0x7e): undefined reference to `tbb::internal::allocate_child_proxy::allocate(unsigned long) const'
spqr_parallel.cpp:(.text._ZN10spqr_zippyISt7complexIdEE7executeEv[_ZN10spqr_zippyISt7complexIdEE7executeEv]+0xca): undefined reference to `tbb::task::spawn_and_wait_for_all(tbb::task_list&)'
spqr_parallel.o: In function `spqr_zippy<double>::execute()':
spqr_parallel.cpp:(.text._ZN10spqr_zippyIdE7executeEv[_ZN10spqr_zippyIdE7executeEv]+0x7e): undefined reference to `tbb::internal::allocate_child_proxy::allocate(unsigned long) const'
spqr_parallel.cpp:(.text._ZN10spqr_zippyIdE7executeEv[_ZN10spqr_zippyIdE7executeEv]+0xca): undefined reference to `tbb::task::spawn_and_wait_for_all(tbb::task_list&)'
spqr_parallel.o: In function `void spqr_parallel<double>(long, int, spqr_blob<double>*)':
spqr_parallel.cpp:(.text._Z13spqr_parallelIdEvliP9spqr_blobIT_E[_Z13spqr_parallelIdEvliP9spqr_blobIT_E]+0x2e): undefined reference to `tbb::task_scheduler_init::initialize(int, unsigned long)'
spqr_parallel.cpp:(.text._Z13spqr_parallelIdEvliP9spqr_blobIT_E[_Z13spqr_parallelIdEvliP9spqr_blobIT_E]+0x38): undefined reference to `tbb::internal::allocate_root_proxy::allocate(unsigned long)'
spqr_parallel.cpp:(.text._Z13spqr_parallelIdEvliP9spqr_blobIT_E[_Z13spqr_parallelIdEvliP9spqr_blobIT_E]+0x77): undefined reference to `tbb::task_scheduler_init::terminate()'
spqr_parallel.cpp:(.text._Z13spqr_parallelIdEvliP9spqr_blobIT_E[_Z13spqr_parallelIdEvliP9spqr_blobIT_E]+0x92): undefined reference to `tbb::task_scheduler_init::terminate()'
spqr_parallel.o: In function `void spqr_parallel<std::complex<double> >(long, int, spqr_blob<std::complex<double> >*)':
spqr_parallel.cpp:(.text._Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E[_Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E]+0x2e): undefined reference to `tbb::task_scheduler_init::initialize(int, unsigned long)'
spqr_parallel.cpp:(.text._Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E[_Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E]+0x38): undefined reference to `tbb::internal::allocate_root_proxy::allocate(unsigned long)'
spqr_parallel.cpp:(.text._Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E[_Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E]+0x77): undefined reference to `tbb::task_scheduler_init::terminate()'
spqr_parallel.cpp:(.text._Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E[_Z13spqr_parallelISt7complexIdEEvliP9spqr_blobIT_E]+0x92): undefined reference to `tbb::task_scheduler_init::terminate()'
spqr_parallel.o:(.data.rel.ro._ZTI10spqr_zippyIdE[_ZTI10spqr_zippyIdE]+0x10): undefined reference to `typeinfo for tbb::task'
spqr_parallel.o:(.data.rel.ro._ZTI10spqr_zippyISt7complexIdEE[_ZTI10spqr_zippyISt7complexIdEE]+0x10): undefined reference to `typeinfo for tbb::task'
spqr_parallel.o:(.data.rel.ro._ZTV10spqr_zippyIdE[_ZTV10spqr_zippyIdE]+0x28): undefined reference to `tbb::task::note_affinity(unsigned short)'
spqr_parallel.o:(.data.rel.ro._ZTV10spqr_zippyISt7complexIdEE[_ZTV10spqr_zippyISt7complexIdEE]+0x28): undefined reference to `tbb::task::note_affinity(unsigned short)'
collect2: error: ld returned 1 exit status
```

jakirkham commented:
Looks like the OpenMP fix is working.

Is it possible the wrong compiler was getting picked up on Ubuntu 14.04? There are complaints like undefined reference to `vtable`, which look a bit strange.

grlee77 commented:
I think the last two commits have addressed the remaining comments.

I can squash a bunch of the commits if you want a cleaner history here.

I see that you already pinned `cvxopt` to use v4.4.* so merging this shouldn't break anything.  Then we can test the `scikit-umfpack` PR over at staged-recipes.

jakirkham commented:
LGTM. Thanks @grlee77. Let's put it through its paces. 🏇
